sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -673,4 +786,81 @@ private void removeOne(String filename) throws IOException {
       }
     }
   }
+
+  /**
+   * Implementations create instances of {@link OutputWrapper} used by
+   * {@link DecoratedFileSink} and related classes to allow <em>decorating</em>, or otherwise
+   * transforming, the raw data that would normally be written directly to the {@link OutputStream}
+   * passed into {@link OutputWrapperFactory#create(OutputStream)}.
+   */
+  public interface OutputWrapperFactory extends Serializable {
+    /**
+     * @param channel the {@link WritableByteChannel} to wrap and allow for decorated, or otherwise
+     *        transformed, data to
+     * @return the {@link OutputWrapper} decorating, or otherwise transforming, the raw data
+     *         before writing it to {@code out}
+     * @throws IOException
+     */
+    public OutputWrapper create(WritableByteChannel channel) throws IOException;
```

Comment:
Why don't the compression factories just return `WritableByteChannel`s themselves? Why are the extra hooks from `OutputWrapper` necessary?